### DISCUSSION TITLE
play 404 default error handler

### DISCUSSION THREAD

**Discussion by abhinandc3** [playframework/playframework#11696] [Q&A] [answered]
does anybody know the error handler for 404 

I am trying to understand how it works internally. for 2.5 
I added a breakpoint to the methods in class 
https://www.playframework.com/documentation/2.5.x/api/java/play/http/DefaultHttpErrorHandler.html
its not getting executed

can someone pls guide what method is getting invoked by default for 404 methods

**Comment by mkurz** [comment, marked as answer]
I can't speak for Play 2.5 because we don't support that anymore and I am not willing to take a look how it worked back then. But as of 2.8.x you find more information in the here (config `play.http.errorHandler`), maybe you need to set the breakpoint in the class `play.api.http.DefaultHttpErrorHandler` (note the `api`)?
https://github.com/playframework/playframework/blob/6aa434ba8a97b287e73a331d12a3c84281b8e80c/core/play/src/main/resources/reference.conf#L49-L57

**Reply by abhinandc3** [reply]
thanks this worked.
one question
DefaultHttpErrorHandler (scala method) is getting invoked
any idea why the java DefaultHttpErrorHandler (without the `api`) is not getting invoked
the app is built on java

**Reply by mkurz** [reply]
Not sure how that was implemented in 2.5.x, but it seems meanwhile implemention details changed:
Actually no matter if a scala or java project, the `play.http.errorHandler` config can either be a class extending `play.http.HttpErrorHandler` or `play.api.http.HttpErrorHandler` (that's also what the comments above says :wink:). It does not matter which of those interface/trait you use. It also does not matter if you inject an subclass of `play.http.HttpErrorHandler` or of `play.api.http.HttpErrorHandler` somewhere in your code, play takes care of providing you the correct handler by converting it to a java or scala one with special adapter and delegate classes.
So by default `play.api.http.DefaultHttpErrorHandler` is set because like said Play can handle both, not matter if a java or scala project.
`play.http.DefaultHttpErrorHandler` still exists to make it easier for Java devs to extend it for customization purposes.
That means, if you want, you can just set `play.http.errorHandler = play.http.DefaultHttpErrorHandler` and everything should work like before and then you can attach the debugger to that java class if that is what you want, to keep it "pure java".
The default setting of using the scala (api) one does not really matter in the end since the they both behave exactly the same: You can compare the two implementations
* [`play.http.DefaultHttpErrorHandler`](https://github.com/playframework/playframework/blob/6aa434ba8a97b287e73a331d12a3c84281b8e80c/core/play/src/main/java/play/http/DefaultHttpErrorHandler.java#L25-L286)
* [`play.api.http.DefaultHttpErrorHandler`](https://github.com/playframework/playframework/blob/6aa434ba8a97b287e73a331d12a3c84281b8e80c/core/play/src/main/scala/play/api/http/HttpErrorHandler.scala#L130-L341)

In theory we (Play) could set `play.http.errorHandler = play.http.DefaultHttpErrorHandler` in the [play-java `reference.conf`](https://github.com/playframework/playframework/blob/main/core/play-java/src/main/resources/reference.conf) so for java projects that one will be used... (In theorie, because having the same config key `play.http.errorHandler` in two reference.conf files doesn't work since it's a gamble which one gets loaded first, so in real life we would need to use a `reference-overrides.conf`). But I don't think this is worth it.
